### PR TITLE
Switchboard

### PR DESCRIPTION
**Pull request by nope-finance**
wip

### CODE REVIEW

**Comment by nope-finance** [comment]
ok should be ready now i think

**Comment by nope-finance** [comment]
test

**Comment by nope-finance** [comment]
test

**Comment by nope-finance** [comment]
test

File: token-lending/program/Cargo.toml [hunk 1]
```diff
@@ -18,8 +18,9 @@ num-derive = "0.3"
 num-traits = "0.2"
 solana-program = "1.7.3"
 spl-token = { path = "../../token/program", features = [ "no-entrypoint" ] }
+switchboard-program = "0.1.45"
 thiserror = "1.0"
-uint = "0.8"
+uint = "0.9.0"
```
Comment (DaSichuan) [inline token-lending/program/Cargo.toml:23]: What is this?

File: token-lending/program/src/processor.rs [hunk 2]
```diff
@@ -406,20 +412,27 @@ fn process_init_reserve(
...
         return Err(LendingError::InvalidAccountInput.into());
     }
 
-    reserve.liquidity.market_price = get_pyth_price(reserve_liquidity_oracle_info, clock)?;
+    if &reserve.liquidity.switchboard_oracle_pubkey != switchboard_feed_info.key {
+        msg!("Lending market quote currency does not match the oracle quote currency");
+        return Err(LendingError::InvalidOracleConfig.into());
+    }
+
+    // reserve.liquidity.market_price = get_pyth_price(pyth_price_info, clock)?;
```
Comment (DaSichuan) [inline token-lending/program/src/processor.rs:434]: Why keep these around?

File: token-lending/program/src/processor.rs [hunk 3]
```diff
@@ -1833,6 +1846,20 @@ fn get_pyth_product_quote_currency(pyth_product: &pyth::Product) -> Result<[u8;
     Err(LendingError::InvalidOracleConfig.into())
 }
 
+fn get_price(
+    switchboard_feed_info: &AccountInfo,
+    pyth_price_account_info: &AccountInfo,
+    clock: &Clock,
+) -> Result<Decimal, ProgramError> {
+    let pyth_price = get_pyth_price(pyth_price_account_info, clock).unwrap_or_default();
+    if pyth_price != Decimal::from(0u64) {
```
Comment (DaSichuan) [inline token-lending/program/src/processor.rs:1855]: Decimal::zero()

File: token-lending/program/src/processor.rs [hunk 4]
```diff
@@ -1833,6 +1846,20 @@ fn get_pyth_product_quote_currency(pyth_product: &pyth::Product) -> Result<[u8;
...
+        return Ok(pyth_price);
+    }
+    let switchboard_price = get_switchboard_price(switchboard_feed_info, clock)?;
+
+    Ok(switchboard_price)
```
Comment (DaSichuan) [inline token-lending/program/src/processor.rs:1860]: You can combine these to just `get_switchboard_price(switchboard_feed_info, clock)`
(no question mark or semicolon)

File: token-lending/program/src/processor.rs [hunk 5]
```diff
@@ -1884,6 +1911,46 @@ fn get_pyth_price(pyth_price_info: &AccountInfo, clock: &Clock) -> Result<Decima
...
+    let price_float = round_result.result.unwrap_or(0.0);
+
+    // we just do this so we can parse coins with low usd value
+    let price_quotient = 10u64.pow(9);
+
+    let price = ((price_quotient as f64) * price_float).round() as u128;
+
+    let decimal_price = Decimal::from(price).try_div(price_quotient)?;
+
+    Ok(decimal_price)
```
Comment (DaSichuan) [inline token-lending/program/src/processor.rs:1951]: Same thing here
`Decimal::from(price).try_div(price_quotient)`

File: token-lending/program/src/processor.rs [hunk 6]
```diff
@@ -1884,6 +1911,46 @@ fn get_pyth_price(pyth_price_info: &AccountInfo, clock: &Clock) -> Result<Decima
...
+        msg!("Switchboard oracle price is stale");
+        return Err(LendingError::InvalidOracleConfig.into());
+    }
+
+    let price_float = round_result.result.unwrap_or(0.0);
+
+    // we just do this so we can parse coins with low usd value
+    let price_quotient = 10u64.pow(9);
+
+    let price = ((price_quotient as f64) * price_float).round() as u128;
```
Comment (DaSichuan) [inline token-lending/program/src/processor.rs:1947]: Should we do some safe multiplication here?

**Review by DaSichuan — CHANGES_REQUESTED** [review]

File: token-lending/program/Cargo.toml (hunk 1 above)
Comment (nope-finance) [reply token-lending/program/Cargo.toml:23]: just some lib for larger int arithmetic? might not need this was getting a warning or something before i think is why i changed it

File: token-lending/program/src/processor.rs (hunk 2 above)
Comment (nope-finance) [reply token-lending/program/src/processor.rs:434]: 4got 2 delete lol

File: token-lending/program/src/processor.rs (hunk 5 above)
Comment (nope-finance) [reply token-lending/program/src/processor.rs:1951]: yeh i guess i was just being more explicit/ if add a 3rd thing but ill just do it that way sure

Comment (DaSichuan) [reply token-lending/program/src/processor.rs:1951]: Yeah i guess it doesn't really matter, but i guess its shorter or something.

File: token-lending/program/src/processor.rs (hunk 6 above)
Comment (nope-finance) [reply token-lending/program/src/processor.rs:1947]: we could but the multiplication isn't the unsafe part the conversion to u128 is

Comment (nope-finance) [reply token-lending/program/src/processor.rs:1944]: added comment flagging this to maybe change later

**Review by DaSichuan — APPROVED** [review]

**Pull request merged**